### PR TITLE
PR from RamR router dyn config

### PR DESCRIPTION
**Pull request by bfallonf**
Pulled #10926 because errors and made my own including some edits.

@ramr Can I get an ack I haven't changed anything too drastically and that this will still work as intended?

### CODE REVIEW

File: install_config/router/default_haproxy_router.adoc [hunk 1]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
 
+[[using-the-dynamic-configuration-manager]]
+== Using the Dynamic Configuration Manager
+
+You can configure the HAProxy router to support the dynamic configuration
+manager.
+
+The dynamic configuration manager brings certain
+types of routes online without requiring any HAProxy reload downtime. It handles any route and endpoint life
+cyle events such as route and endpoint `addition|deletion|update`.
```
Comment (ramr) [inline install_config/router/default_haproxy_router.adoc]: typo `cyle`

File: install_config/router/default_haproxy_router.adoc [hunk 2]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+====
+
+The following procedure assumes you create three routes: `reencrypt-blueprint`, `annotated-edge-blueprint`, and `annotated-unsecured-blueprint`.
+
+.Procedure
+
+. Create a new project:
++
+----
+$ oc new-project project_name
```
Comment (ramr) [inline install_config/router/default_haproxy_router.adoc]: use `namespace_name` instead of `project_name` to match the examples below.

**Review by ramr — APPROVED** [review]
minor comments/typos but otherwise 
/lgtm

**Comment by bfallonf** [comment]
Thanks @ramr . I've fixed those.

@openshift/team-documentation PTAL

File: architecture/topics/router_environment_variables.adoc [hunk 3]
```diff
@@ -75,6 +75,12 @@ endif::[]
 |`ROUTER_DISABLE_NAMESPACE_OWNERSHIP_CHECK` |  | Set to `true` to relax the namespace ownership policy.
 |`ROUTER_STRICT_SNI` |  | xref:strict-sni[strict-sni]
 |`ROUTER_CIPHERS` | intermediate  | Specify the set of xref:ciphers[ciphers] supported by bind.
+|`ROUTER_HAPROXY_CONFIG_MANAGER` | | When set to `true` or `TRUE`, enables a dynamic config manager with haproxy. This allows certain types of routes to be managed dynamically by the config manager and reduces the amount of haproxy router reloads.
```
Comment (kalexand-rh) [inline architecture/topics/router_environment_variables.adoc]: I'd s/amount/number

Comment (kalexand-rh) [reply architecture/topics/router_environment_variables.adoc]: Is this value "False" by default?

File: install_config/router/default_haproxy_router.adoc [hunk 4]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+
+Enable the dynamic configuration manager by setting the
+`ROUTER_HAPROXY_CONFIG_MANAGER` environment variable to `true`:
+
+----
+$ oc set env dc/<router_name> ROUTER_HAPROXY_CONFIG_MANAGER='true'
+----
+
+If the dynamic configuration manager is unable to dynamically configure HAProxy,
+it forces a configuration re-write and a HAProxy process reload. For example, if
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: "For example..." is a fragment. Please rewrite.

File: install_config/router/default_haproxy_router.adoc [hunk 5]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+a newly added route contains custom annotations that would cause a drift from
+the standard configuration (custom timeouts) or if the route requires custom TLS
+configuration.
+
+The dynamic configuration internally uses the haproxy socket and configuration
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: s/haproxy/HAProxy

File: install_config/router/default_haproxy_router.adoc [hunk 6]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+API with a pool of pre-allocated routes and back end servers. The pre-allocated
+pool of routes is created using route blueprints. The default set of blueprints
+supports unsecured routes, edge secured routes (without any custom TLS
+configuration), and passthrough routes.
+
+[IMPORTANT]
+====
+`re-encypt` routes require custom TLS configuration information and so are not
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: s/so/

Comment (kalexand-rh) [reply install_config/router/default_haproxy_router.adoc]: Wait, do we really mean "and so are not supported out of the box?" Do you mean something like

```
`re-encypt` routes require custom TLS configuration information, so you cannot use them directly.

You can extend the blueprints that the dynamic configuration manager uses 
by setting the `ROUTER_BLUEPRINT_ROUTE_NAMESPACE` and optionally the
`ROUTER_BLUEPRINT_ROUTE_LABELS` environment variable values. All routes, or the
routes that match the route labels, in the blueprint route namespace are
processed as custom blueprints similar to the default set of blueprints.

To use `re-encypt` routes or routes
that use custom annotations or routes with custom TLS configuration, define them
through blueprints.
```

File: install_config/router/default_haproxy_router.adoc [hunk 7]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+extended by setting the `ROUTER_BLUEPRINT_ROUTE_NAMESPACE` and optionally the
+`ROUTER_BLUEPRINT_ROUTE_LABELS` environment variable values. All routes (or the
+routes matching the route labels) in the blueprint route namespace will be
+processed as custom blueprints similar to the default set of blueprints. This
+enables defining certain types of routes such as `re-encypt` routes or routes
+that use custom annotations or routes with custom TLS configuration as
+blueprints.
+====
+
+The following procedure assumes you create three routes: `reencrypt-blueprint`, `annotated-edge-blueprint`, and `annotated-unsecured-blueprint`.
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: s/The following procedure assumes/In the following procedure,

File: install_config/router/default_haproxy_router.adoc [hunk 8]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+      --cert=/path/to/cert.pem --service=<service> --port=8443
+----
+
+. Label the route:
++
+----
+$ oc label route edge_route_name type=route_label_1
+----
+
+. xref:../../dev_guide/routes.adoc#creating-routes[Create three different routes]. All have the label `type=route_label_1`:
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: Why is this process different than the route that we created two steps ago?

File: install_config/router/default_haproxy_router.adoc [hunk 9]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
++
+----
+$ oc create -f reencrypt-blueprint.yaml
+$ oc create -f annotated-edge-blueprint.yaml
+$ oc create -f annotated-unsecured-blueprint.json
+----
++
+You can also remove a label from a route, which prevents it from being used as a
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: Is this true even if you don't have a the `ROUTER_BLUEPRINT_ROUTE_LABELS` value set?

File: install_config/router/default_haproxy_router.adoc [hunk 10]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+blueprint route. For example, to prevent the `annotated-unsecured-blueprint`
+from being used as a blueprint route:
++
+----
+$ oc label route annotated-unsecured-blueprint type-
+----
+
+. Create a new router:
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: Is this still part of the same process?

File: install_config/router/default_haproxy_router.adoc [hunk 11]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+. Set the environment variables for the new router:
++
+----
+$ oc set env dc/router ROUTER_HAPROXY_CONFIG_MANAGER=true      \
+                       ROUTER_BLUEPRINT_ROUTE_NAMESPACE=namespace_name  \
+                       ROUTER_BLUEPRINT_ROUTE_LABELS="type=route_label_1"
+----
+
+All routes in the namespace or project `namespace_name` with label
+`type=route_label_1` are able to be processed and used as custom blueprints.
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: s/are able to/can be

File: install_config/router/default_haproxy_router.adoc [hunk 12]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+
+Note that you can also add, update, or remove blueprints by managing the routes
+as you would normally in that namespace `namespace_name`. The dynamic
+configuration manager watches for changes to routes in the namespace
+`namespace_name` similar to how the router watches for `routes` and `services`.
+
+The pool sizes of the pre-allocated routes and back end servers can be
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: This should be a separate step. Like, "Optionally, set blueprint management parameters..." then a ul of the parameters and sample command.

File: install_config/router/default_haproxy_router.adoc [hunk 13]
```diff
@@ -1457,6 +1457,129 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+HAProxy process is reloaded. The default is one hour (3600 seconds) or when the
+dynamic configuration manager runs out of pool space, which ever is sooner. The
+`COMMIT_INTERVAL` environment variable controls this setting.
+
+----
+$ oc set env dc/router -c router ROUTER_BLUEPRINT_ROUTE_POOL_SIZE=20  \
+      ROUTER_MAX_DYNAMIC_SERVERS=3 COMMIT_INTERVAL=6h
+----
+
+The above example increases the pool size for each blueprint route to `20`,
```
Comment (kalexand-rh) [inline install_config/router/default_haproxy_router.adoc]: s/The above/This

**Review by kalexand-rh — COMMENTED** [review]
Some suggestions and picks.

**Comment by bfallonf** [comment]
Changed to suggestion. I'll put this to QA. @lihongan PTAL? This is for 3.11.

File: install_config/router/default_haproxy_router.adoc [hunk 14]
```diff
@@ -1457,6 +1457,131 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+
+The dynamic configuration internally uses the HAProxy socket and configuration
+API with a pool of pre-allocated routes and back end servers. The pre-allocated
+pool of routes is created using route blueprints. The default set of blueprints
+supports unsecured routes, edge secured routes without any custom TLS
+configuration, and passthrough routes.
+
+[IMPORTANT]
+====
+`re-encypt` routes require custom TLS configuration information, so extra
```
Comment (lihongan) [inline install_config/router/default_haproxy_router.adoc]: typo? should be `re-encrypt` routes

File: install_config/router/default_haproxy_router.adoc [hunk 15]
```diff
@@ -1457,6 +1457,131 @@ access any pod in the cluster. If isolation is needed in this case, then do not
...
+configuration is needed in order to use them with the dynamic configuration
+manager.
+
+Extend the blueprints that the dynamic configuration manager can use by setting
+the `ROUTER_BLUEPRINT_ROUTE_NAMESPACE` and optionally the
+`ROUTER_BLUEPRINT_ROUTE_LABELS` environment variables. 
+
+All routes, or the routes that match the route labels, in the blueprint route
+namespace are processed as custom blueprints similar to the default set of
+blueprints. This includes `re-encypt` routes or routes that use custom
```
Comment (lihongan) [inline install_config/router/default_haproxy_router.adoc]: same to above, should be `re-encrypt` routes

File: architecture/topics/router_environment_variables.adoc (hunk 3 above)
Comment (ramr) [reply architecture/topics/router_environment_variables.adoc]: Catching up on emails - yes this is false by default.

File: architecture/topics/router_environment_variables.adoc [hunk 16]
```diff
@@ -75,6 +75,12 @@ endif::[]
...
+|`COMMIT_INTERVAL` | 3600 | Specifies how often to commit changes made with a dynamic config manager. This causes the underlying template router implementation to reload the configuration.
+|`ROUTER_BLUEPRINT_ROUTE_NAMESPACE` | | When set, specifies the namespace which contains routes that serve as blueprints for the dynamic configuration manager. This allows the dynamic configuration manager to support custom routes with any custom annotations, certificates or config.
```
[the remark this answers is not in the document]
Comment (ramr) [reply architecture/topics/router_environment_variables.adoc]: _not_ file

s/config/configuration/

File: architecture/topics/router_environment_variables.adoc [hunk 17]
```diff
@@ -75,6 +75,12 @@ endif::[]
...
+|`ROUTER_BLUEPRINT_ROUTE_LABELS` | | When set, specifies a label selector to apply to the routes in the blueprint route namespace. This allows you to specify certain routes (as opposed to all the routes) in a namespace to serve as blueprints for the dynamic configuration manager.
+|`ROUTER_BLUEPRINT_ROUTE_POOL_SIZE` | 10 | Specifies the size of the pre-allocated pool for each route blueprint that is managed by the dynamic configuration manager. This can be overriden on an individual route basis using an annotation `router.openshift.io/pool-size` on any blueprint route.
+|`ROUTER_MAX_DYNAMIC_SERVERS` | 5 | Specifies the maximum number of dynamic servers added to each route for use by the dynamic configuration manager.
```
[the remark this answers is not in the document]
Comment (ramr) [reply architecture/topics/router_environment_variables.adoc]: 5 overflow servers - this is over and above what the route already has in its config.

As an example: if a route points to service `s1` and that has 4 pods backing that service aka there are 4 endpoints, then the haproxy config would normally have 4 servers for the backend associated with the route. 
In the case of the `ROUTER_HAPROXY_CONFIG_MANAGER=true`, that would be 5 additional ones for the overflow servers making a total of 9 (4 for service `s1` + 5 overflow servers based on the `ROUTER_MAX_DYNAMIC_SERVERS` setting).

File: architecture/topics/router_environment_variables.adoc [hunk 18]
```diff
@@ -75,6 +75,12 @@ endif::[]
...
+|`COMMIT_INTERVAL` | 3600 | Specifies how often to commit changes made with a dynamic config manager. This causes the underlying template router implementation to reload the configuration.
+|`ROUTER_BLUEPRINT_ROUTE_NAMESPACE` | | When set, specifies the namespace which contains routes that serve as blueprints for the dynamic configuration manager. This allows the dynamic configuration manager to support custom routes with any custom annotations, certificates or config.
+|`ROUTER_BLUEPRINT_ROUTE_LABELS` | | When set, specifies a label selector to apply to the routes in the blueprint route namespace. This allows you to specify certain routes (as opposed to all the routes) in a namespace to serve as blueprints for the dynamic configuration manager.
+|`ROUTER_BLUEPRINT_ROUTE_POOL_SIZE` | 10 | Specifies the size of the pre-allocated pool for each route blueprint that is managed by the dynamic configuration manager. This can be overriden on an individual route basis using an annotation `router.openshift.io/pool-size` on any blueprint route.
```
[the remark this answers is not in the document]
Comment (ramr) [reply architecture/topics/router_environment_variables.adoc]: The pool size is for each blueprint. 
As an example, if you create a blueprint route called `joni`, then there would be 10 (based on `ROUTER_BLUEPRINT_ROUTE_POOL_SIZE`) pre-allocated routes named 
`joni-1, joni-2, ... joni-10`.

File: install_config/router/default_haproxy_router.adoc (hunk 9 above)
Comment (ramr) [reply install_config/router/default_haproxy_router.adoc]: No, its only true if the `ROUTER_BLUEPRINT_ROUTE_LABELS` is set.

**Comment by vikram-redhat** [comment]
@bmcelvee can you please incorporate this? Most of the work is done, just a few minor changes requested by @lihongan .

**Comment by bmcelvee** [comment]
@lihongan I fixed the typos, but had to open https://github.com/openshift/openshift-docs/pull/12068 to do so. I'll go ahead and close this PR and merge the content in the new PR. 

Thanks!

**Pull request closed without merging**